### ISSUE TITLE
mutex and locked indicator

### ISSUE THREAD

**Issue by staowyze** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#163]
Hi, this might be a dummy question, I try to understand how the mutex is handled in kinesis, e.g.
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/source/Ice/IceAgent.c#L419
basically, when the mutex is acquired, set the `locked` indicator to TRUE, and during clean up, release the mutex if `locked` is still TRUE, however, from my understanding, set/clear `locked` is not atomic and there is no memory barrier around, how could we guarantee there is no race condition?

**Comment by Sean-Der** [comment]
Hey @staowyze! Really excited you are reading the code, I think it is great for code quality to get more eyes on it :)

`locked` will only be accessed from that current thread/function. We do the `locked` pattern to make sure we don't call `MUTEX_UNLOCK` on a mutex that isn't locked.

We should probably make an abstraction that handles that for us. Have the `locked` on the stack has caused some confusion already.  I would love to a scoped lock like C++, but unfortunately nothing portable like that for C.

**Comment by MushMal** [comment]
To iterate on what @Sean-Der said. This is an established pattern used to handle locking/unlocking without a compiler scope support. This is also useful when handling multiple locks in a secure fashion which won't be possible with traditional C++ scoping.

Resolving this issue. Please re-open if you need more info

**Comment by staowyze** [comment]
the problem is `locked` is not atomic, there is no memory barrier, i.e. the CPU could execute instructions out of order, even in a single thread. there could be a scenario that, we lock the mutex, but the CPU doesn't set `locked` to TRUE due to out of order execution.

**Comment by Sean-Der** [comment]
@staowyze Oh I see. 

If this is an issue I see no problem with using atomic get/set every time! We should for sure do a deeper audit. All of the code I have written assumes in order execution.

**Comment by Sean-Der** [comment]
@MushMal reopening, would love to hear your opinion.

**Comment by staowyze** [comment]
Yeah, I mixed atomicity and memory barrier as memory barrier is guaranteed for atomic variable.
I am concerning the following example
```
MUTEX_LOCK(pIceAgent->lock);
locked = TRUE;

CHK(FALSE, retStatus)

CleanUp:

    if (locked) {
        MUTEX_UNLOCK(pIceAgent->lock);
    }
```
If the CPU decides to run `CHK` first without setting `locked`, then appears `pIceAgent->lock` is not released. Hope the computer is smart enough to figure out the dependency.

**Comment by MushMal** [comment]
@staowyze as the code is local, the compiler/platform (including CPU) combination guarantees the order. So, for example, in this case, MUTEX_LOCK and locked = TRUE are macro-instructions which can be re-ordered. However, when running the conditional in the CHK statement, the compiler will guarantee the right view of the execution. This won't be the case if there were multiple threads involved - or, a device/memory mapped region gets updated.

To summarize, as this piece of code is run in a single thread context (doesn't matter whether the OS pre-empts the execution and runs on another physical core - the thread context should capture the locals and raise a full fence) the logical order will be guaranteed.

**Comment by MushMal** [comment]
Resolving for now. Please re-open is you have more questions

**Comment by staowyze** [comment]
thank you for the clarification.
